### PR TITLE
C++: Fix joins in `cpp/using-expired-stack-address`

### PR DESCRIPTION
**Pull request by MathiasVP**
These were highlighted by DCA as problematic joins.

Before:
```ql
Evaluated recursive predicate UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0@f7cfb3wg in 8404ms on iteration 5 (delta size: 298908).
Evaluated relational algebra for predicate UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0@f7cfb3wg on iteration 5 running pipeline order_500000 with tuple counts:
  ...

  243714   ~1%    {3} r15 = JOIN UsingExpiredStackAddress#38a029fe::TFieldAddress#fff#prev_delta WITH UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0#prev ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2
  318964   ~0%    {3} r16 = JOIN r15 WITH Instruction#577b6a83::FieldAddressInstruction::getObjectAddress#dispred#ff_10#join_rhs ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2
   31341   ~7%    {2} r17 = JOIN r16 WITH Instruction#577b6a83::FieldInstruction#ff ON FIRST 2 OUTPUT Lhs.2, Lhs.0

  ...

    35231   ~0%    {4} r29 = JOIN UsingExpiredStackAddress#38a029fe::TConversion#fffff#prev_delta WITH UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0#join_rhs#1 ON FIRST 1 OUTPUT Lhs.2, Lhs.3, Lhs.1, Lhs.4
 71215721   ~3%    {3} r30 = JOIN r29 WITH UsingExpiredStackAddress#38a029fe::uncheckedConversionTypes#fff_120#join_rhs ON FIRST 2 OUTPUT Rhs.2, Lhs.2, Lhs.3
 71215721   ~0%    {4} r31 = JOIN r30 WITH Instruction#577b6a83::UnaryInstruction::getUnary#dispred#ff ON FIRST 1 OUTPUT Lhs.1, Rhs.1, Lhs.2, Lhs.0
    42934   ~2%    {2} r32 = JOIN r31 WITH UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0#prev ON FIRST 2 OUTPUT Lhs.2, Lhs.3

  ...
```
After:
```ql
Evaluated recursive predicate UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0@0a9393wo in 176ms on iteration 5 (delta size: 298908).
Evaluated relational algebra for predicate UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0@0a9393wo on iteration 5 running pipeline order_500000 with tuple counts:
  ...

 243714   ~1%    {3} r9 = JOIN UsingExpiredStackAddress#38a029fe::TFieldAddress#fff#prev_delta WITH UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0#prev ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2
  31341   ~7%    {2} r10 = JOIN r9 WITH UsingExpiredStackAddress#38a029fe::instructionHasFieldAndObject#fff_120#join_rhs ON FIRST 2 OUTPUT Lhs.2, Rhs.2
  
  ...
   
  35231   ~0%    {4} r25 = JOIN UsingExpiredStackAddress#38a029fe::TConversion#fffff#prev_delta WITH UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0#join_rhs#1 ON FIRST 1 OUTPUT Lhs.1, Lhs.2, Lhs.3, Lhs.4
  80392   ~1%    {4} r26 = JOIN r25 WITH UsingExpiredStackAddress#38a029fe::globalAddress#ff#reorder_1_0#prev ON FIRST 1 OUTPUT Rhs.1, Lhs.1, Lhs.2, Lhs.3
  42934   ~2%    {2} r27 = JOIN r26 WITH UsingExpiredStackAddress#38a029fe::Conversion#Instruction#577b6a83::ConvertInstruction#::Using#uncheckedConversionTypes#::hasOperandAndTypes#ffff_1230#join_rhs ON FIRST 3 OUTPUT Lhs.3, Rhs.3

```

### CODE REVIEW

**Comment by MathiasVP** [comment]
@ginsbach would you please double-check that I'm not misusing parameterized modules in any way? 🙏

**Review by ginsbach — COMMENTED** [review]
I have no concerns about the use of parameterised modules here.

**Review by jketema — APPROVED** [review]
I'm happy if DCA is happy.

File: cpp/ql/src/Likely Bugs/Memory Management/UsingExpiredStackAddress.ql [hunk 1]
```diff
@@ -117,25 +137,27 @@ TGlobalAddress globalAddress(Instruction instr) {
     result = TLoad(globalAddress(load.getSourceAddress()))
   )
   or
-  exists(ConvertInstruction convert, Type fromType, Type toType | instr = convert |
-    uncheckedConversionTypes(convert, fromType, toType) and
-    result = TConversion("unchecked", globalAddress(convert.getUnary()), fromType, toType)
+  exists(Type fromType, Type toType, Instruction unary |
```
Comment (andersfugmann) [inline cpp/ql/src/Likely Bugs/Memory Management/UsingExpiredStackAddress.ql:140]: Nit: I wonder is we could lift all the logic into the new module. So essentially the three blocks would read:

```codeql
Conversion<ConvertInstruction>::Using<uncheckedConversionTypes/3>::existInstruction("unchecked") or
Conversion<CheckedConvertOrNullInstruction>::Using<checkedConversionTypes/3>::existInstruction("checked") or
Conversion<InheritanceConversionInstruction>::Using<inheritanceConversionTypes/3>::existInstruction("inheritance")
```

Or something simillar, now we are already moving parts of the logic into a module parameterized with both the instruction type and the predicate to be used.

**Review by andersfugmann — APPROVED** [review]
LGTM

File: cpp/ql/src/Likely Bugs/Memory Management/UsingExpiredStackAddress.ql (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/Likely Bugs/Memory Management/UsingExpiredStackAddress.ql:140]: We probably could. Note that we really only introduced this module to force a specific join order. So if we move more things into the module then we'll probably reintroduce the bad join order.

**Comment by MathiasVP** [comment]
DCA looks good: the offending predicate has disappeared from the join-order reports 🎉.

**Pull request merged**